File: jehyeok/두큐합같게만들기.py
from collections import deque

queue1 = [1, 1]
queue2 = [1]

def solution(queue1, queue2):
    answer = 0

    # 총 합이 홀수면 두 큐를 같게 만들 수 없음
    if (sum(queue1) + sum(queue2)) % 2 == 1:
        return -1
    
    queue1 = deque(queue1)
    queue2 = deque(queue2)

    queue1Sum = sum(queue1)
    queue2Sum = sum(queue2)

    while True:
        if len(queue1) == 0 or len(queue2) == 0:
            answer = -1
            break

        # queue1 과 queue2가 서로 뒤바뀌었을때
        if 2 * (len(queue1) + len(queue2)) < answer:
            answer = -1
            break

        if queue1Sum == queue2Sum:
            break

        if queue1Sum > queue2Sum:
            v = queue1.popleft()
            queue2.append(v)
            queue1Sum -= v
            queue2Sum += v
        else:
            v = queue2.popleft()
            queue1.append(v)
            queue2Sum -= v
            queue1Sum += v

        answer += 1

    return answer

print(solution(queue1, queue2))

File: younghyo/두 큐 합 같게 만들기.java

// 각 큐의 합
// 큐ㅠ...

import java.util.*;
class Solution {
    public int solution(int[] queue1, int[] queue2) {
        int answer = 0;

        int n = queue1.length;

        long[] merge = new long[n * 2]; // 하나의 배열로 합치

        for (int i = 0; i < n; i++) {
            merge[i] = queue1[i];
        }

        for (int i = 0; i < n; i++) {
            merge[n + i] = queue2[i];
        }

        long q1Sum = Arrays.stream(queue1).sum();
        long q2Sum = Arrays.stream(queue2).sum();


        if ((q1Sum + q2Sum) % 2 != 0) {
            return -1;
        }

        long result = (q1Sum + q2Sum) / 2;

        Arrays.sort(queue1);
        Arrays.sort(queue2);

        int i = 0, j = 0;

        while (i < n && j < n) {
            if (q1Sum + queue1[i] <= result) {
                q1Sum += queue1[i++];
            } else {
                q2Sum += queue1[i++] - result +  q2Sum;
                answer++;
            }

            if (q2Sum + queue2[j] <= result) {
                q2Sum += queue2[j++];
            } else {
                q1Sum  += queue2[j++] - result + q1Sum ;
                answer++;
            }
        }

        return answer;
    }

}